Switch from 'kintone-java-sdk' to 'kintone-java-client'

abetomo opened the pull request:
Change to use [kintone-java-client](https://github.com/kintone/kintone-java-client) in reference to https://github.com/trocco-io/embulk-output-kintone/pull/6#issuecomment-671660006.

## Issues that arise

Configuration:

If it is `mode: UPDATE`, it works as expected.
If it is `mdoe: update`, it is an error.
<details>
<summary>I get the following error.</summary>

```
org.embulk.exec.PartialExecutionException: org.embulk.config.ConfigException: java.lang.IllegalArgumentException: Can not construct instance of org.embulk.output.kintone.KintoneMode from String value 'update': value not one of declared Enum instance names: [INSERT, UPSERT, UPDATE]
 at [Source: N/A; line: -1, column: -1]
	at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(BulkLoader.java:340)
	at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:566)
	at org.embulk.exec.BulkLoader.access$000(BulkLoader.java:35)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:353)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:350)
	at org.embulk.spi.Exec.doWith(Exec.java:22)
	at org.embulk.exec.BulkLoader.run(BulkLoader.java:350)
	at org.embulk.EmbulkEmbed.run(EmbulkEmbed.java:242)
	at org.embulk.EmbulkRunner.runInternal(EmbulkRunner.java:291)
	at org.embulk.EmbulkRunner.run(EmbulkRunner.java:155)
	at org.embulk.cli.EmbulkRun.runSubcommand(EmbulkRun.java:431)
	at org.embulk.cli.EmbulkRun.run(EmbulkRun.java:90)
	at org.embulk.cli.Main.main(Main.java:64)
	Suppressed: java.lang.NullPointerException
		at org.embulk.exec.BulkLoader.doCleanup(BulkLoader.java:463)
		at org.embulk.exec.BulkLoader$3.run(BulkLoader.java:397)
		at org.embulk.exec.BulkLoader$3.run(BulkLoader.java:394)
		at org.embulk.spi.Exec.doWith(Exec.java:22)
		at org.embulk.exec.BulkLoader.cleanup(BulkLoader.java:394)
		at org.embulk.EmbulkEmbed.run(EmbulkEmbed.java:245)
		... 5 more
Caused by: org.embulk.config.ConfigException: java.lang.IllegalArgumentException: Can not construct instance of org.embulk.output.kintone.KintoneMode from String value 'update': value not one of declared Enum instance names: [INSERT, UPSERT, UPDATE]
 at [Source: N/A; line: -1, column: -1]
	at org.embulk.config.ModelManager.readObjectWithConfigSerDe(ModelManager.java:78)
	at org.embulk.config.DataSourceImpl.loadConfig(DataSourceImpl.java:197)
	at org.embulk.output.kintone.KintoneOutputPlugin.transaction(KintoneOutputPlugin.java:23)
	at org.embulk.exec.BulkLoader$4$1$1.transaction(BulkLoader.java:521)
	at org.embulk.exec.LocalExecutorPlugin.transaction(LocalExecutorPlugin.java:50)
	at org.embulk.exec.BulkLoader$4$1.run(BulkLoader.java:516)
	at org.embulk.spi.util.Filters$RecursiveControl.transaction(Filters.java:84)
	at org.embulk.spi.util.Filters.transaction(Filters.java:42)
	at org.embulk.exec.BulkLoader$4.run(BulkLoader.java:511)
	at org.embulk.spi.FileInputRunner$RunnerControl$1$1.run(FileInputRunner.java:112)
	at org.embulk.standards.CsvParserPlugin.transaction(CsvParserPlugin.java:237)
	at org.embulk.spi.FileInputRunner$RunnerControl$1.run(FileInputRunner.java:107)
	at org.embulk.spi.util.Decoders$RecursiveControl.transaction(Decoders.java:68)
	at org.embulk.spi.util.Decoders.transaction(Decoders.java:29)
	at org.embulk.spi.FileInputRunner$RunnerControl.run(FileInputRunner.java:105)
	at org.embulk.standards.LocalFileInputPlugin.resume(LocalFileInputPlugin.java:79)
	at org.embulk.standards.LocalFileInputPlugin.transaction(LocalFileInputPlugin.java:72)
	at org.embulk.spi.FileInputRunner.transaction(FileInputRunner.java:62)
	at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:507)
	... 11 more
Caused by: java.lang.IllegalArgumentException: Can not construct instance of org.embulk.output.kintone.KintoneMode from String value 'update': value not one of declared Enum instance names: [INSERT, UPSERT, UPDATE]
 at [Source: N/A; line: -1, column: -1]
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3459)
	at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:3400)
	at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:3388)
	at org.embulk.config.TaskSerDe$TaskDeserializer.deserialize(TaskSerDe.java:152)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3708)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2005)
	at org.embulk.config.ModelManager.readObjectWithConfigSerDe(ModelManager.java:73)
	... 29 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Can not construct instance of org.embulk.output.kintone.KintoneMode from String value 'update': value not one of declared Enum instance names: [INSERT, UPSERT, UPDATE]
 at [Source: N/A; line: -1, column: -1]
	at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:55)
	at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:907)
	at com.fasterxml.jackson.databind.deser.std.EnumDeserializer._deserializeAltString(EnumDeserializer.java:130)
	at com.fasterxml.jackson.databind.deser.std.EnumDeserializer.deserialize(EnumDeserializer.java:84)
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:3454)
	... 35 more

Error: org.embulk.config.ConfigException: java.lang.IllegalArgumentException: Can not construct instance of org.embulk.output.kintone.KintoneMode from String value 'update': value not one of declared Enum instance names: [INSERT, UPSERT, UPDATE]
 at [Source: N/A; line: -1, column: -1]
```
</details>
I tried several solutions, but none of them worked. 

Could you give me some advice about this problem?

abetomo committed:
Switch from 'kintone-java-sdk' to 'kintone-java-client'

abetomo committed:
Fix setting of timestamp value

giwa commented:
I do not come up good solution for this right now. This problem is related to Jackson incompatibility. The current approach is mapping yaml to EUM in Java. We can switch it to String and add some validation.

> If it is mode: UPDATE, it works as expected.
> If it is mdoe: update, it is an error.

abetomo committed:
Change `task.getMode()` to return String

abetomo commented:
Thanks for the comment.
I changed it to be handled by String.

giwa commented on README.md:
```diff
@@ -5,7 +5,7 @@
 
 kintone output plugin for Embulk stores app records from kintone.
-embulk 0.9 is only supported due to the dependency of kintone-java-sdk 0.4.0, which requires java 8
+embulk 0.9 is only supported due to the dependency of [kintone-java-client](https://github.com/kintone/kintone-java-client), which requires java 8.
```
You can delete this comment. The motivation for this comment was to explain why we didn't  choose the latest kintone-java-sdk. We don't need to keep this comment since kintone-java-client supports Java from 8

abetomo replied:
I see.
I deleted it.

giwa reviewed (commented):
Cool:) Did you test transfer some data with this change and paste log, please? I want to make sure everything works fine.

abetomo committed:
Remove description of dependencies from README

https://github.com/trocco-io/embulk-output-kintone/pull/8#discussion_r470475139

```
You can delete this comment. The motivation for this comment was to explain why we didn't choose the latest kintone-java-sdk. We don't need to keep this comment since kintone-java-client supports Java from 8
```

abetomo commented:
The log is as follows.
Please let us know if there is any other information you need.

```yaml
in:
  type: file
  path_prefix: ./csv/
  parser:
    charset: UTF-8
    newline: CRLF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '"'
    null_string: 'NULL'
    columns:
    - {name: date, type: timestamp, format: '%Y-%m-%d'}
    - {name: text, type: string}
    - {name: datetime, type: timestamp, format: '%Y-%m-%d %H:%M:%S%z'}
    - {name: num, type: long}
out:
  type: kintone
  domain: ".cybozu.com"
  token: xxx
  app_id: N
  mode: update
  column_options:
    date: {field_code: "date", type: "DATE"}
    text: {field_code: "text1", type: "SINGLE_LINE_TEXT"}
    datetime: {field_code: "datetime", type: "DATETIME"}
    num: {field_code: "num", type: "NUMBER", update_key: true}
```

```
% cat csv/test.csv
2002-02-21,text999,1999-11-12 10:10:10+09:00,3
4442-08-31,text888,2002-12-13 11:11:11+09:00,4
```

```
% ./gradlew clean build && ./gradlew classpath && embulk run -I ./lib ./config.yml

BUILD SUCCESSFUL in 1s
7 actionable tasks: 7 executed

BUILD SUCCESSFUL in 8s
4 actionable tasks: 2 executed, 2 up-to-date
2020-08-14 08:24:58.926 +0000: Embulk v0.9.23
2020-08-14 08:24:59.676 +0000 [WARN] (main): DEPRECATION: JRuby org.jruby.embed.ScriptingContainer is directly injected.
2020-08-14 08:25:01.545 +0000 [INFO] (main): Gem's home and path are set by default: "/home/vagrant/.embulk/lib/gems"
2020-08-14 08:25:02.457 +0000 [INFO] (main): Started Embulk v0.9.23
2020-08-14 08:25:02.578 +0000 [INFO] (0001:transaction): Loaded plugin embulk/output/kintone from a load path
2020-08-14 08:25:02.614 +0000 [INFO] (0001:transaction): Listing local files at directory './csv' filtering filename by prefix ''
2020-08-14 08:25:02.616 +0000 [INFO] (0001:transaction): "follow_symlinks" is set false. Note that symbolic links to directories are skipped.
2020-08-14 08:25:02.619 +0000 [INFO] (0001:transaction): Loading files [./csv/test.csv]
2020-08-14 08:25:02.686 +0000 [INFO] (0001:transaction): Using local thread executor with max_threads=4 / output tasks 2 = input tasks 1 * 2
2020-08-14 08:25:02.696 +0000 [INFO] (0001:transaction): {done:  0 / 1, running: 0}
2020-08-14 08:25:03.528 +0000 [INFO] (0001:transaction): {done:  1 / 1, running: 0}
2020-08-14 08:25:03.533 +0000 [INFO] (main): Committed.
2020-08-14 08:25:03.533 +0000 [INFO] (main): Next config diff: {"in":{"last_path":"./csv/test.csv"},"out":{}}
```

giwa commented:
@abetomo I gonna merge this. Thanks!
